### PR TITLE
Kl weighting by iterate

### LINKED ISSUE #507: KL warmup should be timed to the number of iterations, not epochs
For very large datasets the current implementation can result in too much warmup.

### PR DESCRIPTION
**Pull request by PierreBoyeau**
Fixes #507

### CODE REVIEW

File: scvi/inference/inference.py [hunk 1]
```diff
@@ -42,15 +46,26 @@ class UnsupervisedTrainer(Trainer):
...
+        n_iter_kl_warmup: int = None,
+        n_epochs_kl_warmup: int = 400,
+        normalize_loss: bool = None,
         **kwargs
     ):
         super().__init__(model, gene_dataset, **kwargs)
+
+        # Set up number of warmup iterations
+        n_reference_cells = min(10000, gene_dataset.nb_cells)
+        if n_iter_kl_warmup is not None:
```
Comment (adamgayoso) [inline scvi/inference/inference.py]: I'm not sure I understand this default. if you set a not None value that should be what is used. I was only suggesting 400 * 10000/batch_size as the default in the function signature.

File: scvi/inference/trainer.py [hunk 2]
```diff
@@ -153,13 +161,14 @@ def train(
         ):
             self.on_epoch_begin()
             for tensors_list in self.data_loaders_loop():
+                self.on_iteration_begin()
                 if tensors_list[0][0].shape[0] < 3:
                     continue
-                loss = self.loss(*tensors_list)
-                self.check_training_status(loss, max_nans)
+                self.current_loss = loss = self.loss(*tensors_list)
```
Comment (adamgayoso) [inline scvi/inference/trainer.py]: what's the implication of the training status now being checked AFTER the backward call?

Comment (PierreBoyeau) [reply scvi/inference/trainer.py]: I don't think it has any impact: 

```
import torch

param = torch.tensor([10.], requires_grad=True)
optim = torch.optim.SGD([param], lr=0.1)
loss = param ** 2
print(loss)
loss.backward()
print(loss)

```

File: scvi/inference/inference.py (hunk 1 above)
Comment (PierreBoyeau) [reply scvi/inference/inference.py]: The motivation for `n_reference_cells = min(10000, gene_dataset.nb_cells)` was to take into account the dataset size to set up the threshold iteration number.

Comment (adamgayoso) [reply scvi/inference/inference.py]: It's still a bit weird to me that you put a number for this and then it gets changed. So if I set this to be 100 iters, and my dataset is very large, then the trainer changes it to (10000/128) * 100?

I'm suggesting that if you set it to 100 iters that's all it should do. The default I'm suggesting is so that the performance of scVI doesn't change very much if doing iter warmup vs epoch warmup.

File: scvi/inference/trainer.py (hunk 2 above)
Comment (adamgayoso) [reply scvi/inference/trainer.py]: Sorry got confused with the iteration end and the epoch end and the early stopping stuff.

File: scvi/inference/inference.py (hunk 1 above)
Comment (PierreBoyeau) [reply scvi/inference/inference.py]: Ok I think I get it!
